### PR TITLE
enhancement/read-only-on-get

### PR DESCRIPTION
**Pull request by decanus**

### CODE REVIEW

**Review by mandrigin — APPROVED** [review]
yep, that is exactly what I meant! :+1:

File: state.go [hunk 1]
```diff
@@ -16,7 +16,19 @@ type syncState struct {
 	state map[GroupID]map[MessageID]map[PeerId]state
 }
 
+
 func (s syncState) Get(group GroupID, id MessageID, sender PeerId) state {
+	s.RLock()
+	defer s.RUnlock()
+
+	if _, ok := s.state[group][id][sender]; !ok {
```
Comment (mandrigin) [inline state.go]: well, then you theoreticaly can even do 
```
state, _ := s.state[group][id][sender]
return state
```

Comment (decanus) [reply state.go]: Would state not be nil @mandrigin?

Comment (mandrigin) [reply state.go]: it it is not a pointer, it shoudn't be it will be a default value of the struct

**Comment by decanus** [comment]
@mandrigin running this code seems to run into a deadlock at one point, need to investigate further.

File: state.go (hunk 1 above)
Comment (mandrigin) [reply state.go]: see https://play.golang.org/p/mSTMq4zvLeV

**Comment by mandrigin** [comment]
It’s unlikely there, this code is pretty straightforward, but if you call Get or Set inside Map, that might bring trouble.

**Pull request merged**